### PR TITLE
Update Test Library

### LINKED ISSUE #78: [Asyncio] Test suite is failing
While investigating what it would take to enable 3.7 compatibility, I discovered the test suite is not currently in a passing state and, therefore, it is not possible to reliably prove any fix would take effect without introducing bugs / breaking behaviors. 

This is very worrying, as my organization is building a IRC bot using this library that cannot blow up randomly do to an undetected bug in the underlying IRC infrastructure. As the tests are not passing, the stability of this project cannot be reliably proven. While Pydle appears to function as it is documented based on integration tests, It may prove difficult to both maintain the project and upgrade it without passing tests. 

Environment: 
Windows 10 x64, Linux
python x64 v `3.6.3` and `3.6.6`

Test summery:
```
================================================= test session starts ==================================================
platform linux -- Python 3.6.6, pytest-3.6.3, py-1.5.4, pluggy-0.6.0
rootdir: /pydle, inifile: tox.ini
collected 53 items

tests/test__fixtures.py ...F                                                                                     [  7%]
tests/test__mocks.py FFF.........                                                                                [ 30%]
tests/test_client.py .FF.F.FFFFFF                                                                                [ 52%]
tests/test_client_channels.py ......                                                                             [ 64%]
tests/test_client_users.py ..............                                                                        [ 90%]
tests/test_featurize.py .....                                                                                    [100%]

```
Full error output: [gist](https://gist.github.com/theunkn0wn1/3f54475ff973db20134b0ab0772ca122)

looks like there is an issue somewhere in `MockClient._connect`, however I have been unable to narrow it down. 

My apologies for wasting your time  should this already be a known issue, for I am unable to find any reference to tests in recent issue tickets.

### PR DESCRIPTION
**Pull request by Rixxan**
The majority of the test library hasn't been touched in around 8 years. As such, most of the tests were not properly set for async testing and were not functioning properly. 

This update properly assigns async functionality to a number of the tests and makes efforts to modernize the bulk of the test library that already exists. These updates bring all 59 tests included in the package to passing status in Python 3.5, 3.8, and 3.9

The test library still needs serious work, however, and is not currently a comprehensive test package at this time. It is at the limits of my ability to write tests, however, and additional tests should be written.

Resolves #78

### CODE REVIEW

File: tests/fixtures.py [hunk 1]
```diff
@@ -9,11 +9,13 @@ def with_client(*features, connected=True, **options):
...
-        def run():
+        async def run():
             server = MockServer()
-            client = with_client.classes[features]('TestcaseRunner', mock_server=server, **options)
+            client = with_client.classes[features](
+                "TestcaseRunner", mock_server=server, **options
+            )
             if connected:
-                client.connect('mock://local', 1337, eventloop=MockEventLoop())
+                await client.connect("mock://local", 1337, eventloop=MockEventLoop())
```
Comment (theunkn0wn1) [inline tests/fixtures.py:18]: We should get rid of the mock event loop, its a relic of a bygone era.

File: tests/mocks.py [hunk 2]
```diff
@@ -104,17 +110,17 @@ def off(self, *args, **kwargs):
...
         self._mock_server.connection = self
         self._mock_connected = True
 
-    def disconnect(self, *args, **kwargs):
+    async def disconnect(self, *args, **kwargs):
         self._mock_server.connection = None
         self._mock_connected = False
 
 
 class MockEventLoop:
```
Comment (theunkn0wn1) [inline tests/mocks.py:122]: We should dispose of this class, pydle should not be rolling its own event loop in this era.

File: tests/test__mocks.py [hunk 3]
```diff
@@ -91,28 +102,31 @@ def test_mock_eventloop_schedule():
 
     ev.stop()
 
+
 @mark.meta
 @mark.slow
 def test_mock_eventloop_schedule_in():
```
Comment (theunkn0wn1) [inline tests/test__mocks.py:108]: meta tests related to the mock event loop can also go

File: tests/test_client.py [hunk 4]
```diff
@@ -10,47 +10,57 @@
...
+@pytest.mark.asyncio
 @with_client()
-def test_client_reconnect(server, client):
-    client.disconnect(expected=True)
+async def test_client_reconnect(server, client):
+    await client.disconnect(expected=True)
     assert not client.connected
 
-    client.connect(reconnect=True)
+    await client.connect(reconnect=True)
```
Comment (theunkn0wn1) [inline tests/test_client.py:27]: Im not convinced this test is enough, disconnection logic has been a particular trouble spot for this library.

If you reduce the disconnect timer on the mock client and sleep past it, we can add an assertion here that pydle didn't reconnect itself when it wasn't supposed to.

**Review by theunkn0wn1 — APPROVED** [review]
Thanks for this!
While we are here, can we remove the mock event loop?

File: tests/test_client.py (hunk 4 above)
Comment (Rixxan) [reply tests/test_client.py:27]: Ultimately I lack the knowledge of the disconnection logic enough to make that change and have any idea what I'm doing, so I will leave that for a later coder to solve, unfortunately.

Comment (theunkn0wn1) [reply tests/test_client.py:27]: fair enough, problem for a later time.

**Pull request merged**